### PR TITLE
Implementation / integration  of the `Profile payment`/`checkout` #22

### PR DESCRIPTION
**Pull request by fetahokey**
##  Summary
- To satisfied, the requirement of the payment `customerId` attribute added to the `profile` model, 
- When a user navigates to the payment page for the first time, a stripe customer will be created for the authenticated user if `customerId` is not defined yet, 
- Each payment profile (payment method) created will be attached to the stripe customer as a source, 
- On the payment page, the frontend will fetch the payment methods from the stripe server 
- Checkout creates a Payment intent, that will be confirmed on the frontend using a client secret, after successful payment confirmation, the request status will be updated to `paid`
  
- Functionalities:  
  - Payment profiles UI
    - Payment profiles (payment methods) integration 
    - Add a new `payment method`
    - Set default `payment method`
  - Checkout UI
    - Checkout integration with the `booking` page
    - Popup `successful`, `error` message in the `booking` page, after payment confirmation

### Screenshots / Videos :

>Select payment options
>select payment options

>Add new payment method
>add new payment method

>Add second payment method
>add second payment method

>Payment method added
>payment method added

> Select default payment method
>Select default payment method

> successful payment message
>confirm payment

### Information needed to test this feature:
- Mock-data on the database for the `Requests` document
- Setup `private` , `public` stripe key
- Users roles should be set to 'sitter'
- Start the app and navigate 
  - to `/booking`, to test the `checkout`
  - to `/payment`, to test the `payment` profiles

### Limitations with the current functionality:
- Refund
- Payment cancelation
- Payout (exp: sitter)

### Features:
- [x] #34 
- [x] #22
- [ ] Payout 
- [ ] Reinforce security

### CODE REVIEW

File: server/controllers/stripe.js [hunk 1]
```diff
@@ -25,25 +30,190 @@ const getRequestPrice = asyncHandler(async (requestId, res) => {
...
       }
     }
   );
 });
+
+// @route POST /payments/customer
+// @desc create a stripe customer
+// @access Private
+exports.createCustomer = asyncHandler(async (req, res, next) => {
+  const { name, email, phone } = req.body;
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:62]: Validate these and return a `400` if any is missing in the request body.

File: server/controllers/stripe.js [hunk 2]
```diff
@@ -25,25 +30,190 @@ const getRequestPrice = asyncHandler(async (requestId, res) => {
...
+        res.status(500);
+        throw new Error("Internal Server Error");
+      }
+    }
+  );
+});
+
+// @route PATCH /payments/customer
+// @desc update a stripe customer default paymentMethod (default source)
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:87]: Nice validation

File: server/controllers/stripe.js [hunk 3]
```diff
@@ -25,25 +30,190 @@ const getRequestPrice = asyncHandler(async (requestId, res) => {
...
+exports.updateCustomer = asyncHandler(async (req, res, next) => {
+  const { cardId, customerId } = req.body;
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:89]: Validate these and return a `400` if any is missing in the request body.

File: server/controllers/stripe.js [hunk 4]
```diff
@@ -25,25 +30,190 @@ const getRequestPrice = asyncHandler(async (requestId, res) => {
...
+          stripeResponse,
+        });
+      }
+    }
+  );
+});
+
+// @route POST /payments/method/
+// @desc create stripe paymentMethod `token` and bind the token to the current customer
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:109]: :+1:

File: server/controllers/stripe.js [hunk 5]
```diff
@@ -25,25 +30,190 @@ const getRequestPrice = asyncHandler(async (requestId, res) => {
...
+exports.createPaymentMethod = asyncHandler(async (req, res, next) => {
+  const { creditCardNumber, cvcField, expDate, name, customerId } = req.body;
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:111]: Validate these and return a `400` if any is missing in the request body.

File: server/controllers/stripe.js [hunk 6]
```diff
@@ -25,25 +30,190 @@ const getRequestPrice = asyncHandler(async (requestId, res) => {
...
+        );
+      }
+    }
+  );
+});
+
+// @route GET payments/method/:id
+// @desc list payment methods for a given stripe customer
+// @access Private
+// @author @fetahokey
```
Comment (ezeilo-su) [inline server/controllers/stripe.js:152]: :+1:

**Review by ezeilo-su — COMMENTED** [review]
This one is good!

File: server/controllers/stripe.js (hunk 1 above)
Comment (fetahokey) [reply server/controllers/stripe.js:62]: > It already has been validated, check the strips.js in the  route folder, validators they were added as middleware 
 ```
 router.route("/customer").post(protect, validateCreateCustomer, createCustomer);

File: server/controllers/stripe.js (hunk 5 above)
Comment (fetahokey) [reply server/controllers/stripe.js:111]: > It already has been validated, 
please check the strips.js in the  route folder, 
 ```
 const {
  validateStripeRequest,
  validateListPaymentMethods,
  validateCreateCustomer,
  validateUpdateCustomer,
  validateCreatePaymentMethod,
} = require("../validators/stripe");

File: server/controllers/stripe.js (hunk 3 above)
Comment (fetahokey) [reply server/controllers/stripe.js:89]: > It already has been validated, 
please check the strips.js in the  route folder, 
 ```
 const {
  validateStripeRequest,
  validateListPaymentMethods,
  validateCreateCustomer,
  validateUpdateCustomer,
  validateCreatePaymentMethod,
} = require("../validators/stripe");

File: client/src/pages/Payment/CreditCard.tsx [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        defaultProfileAction();
+      }
+    });
+  };
+  return (
+    <>
+      <Box width={'60%'} mr={10}>
+        <Card elevation={1}>
+          <CardHeader
+            avatar={<img src={`./assets/${type}.png`} className={cardBrandLogo} />}
```
Comment (rajivtitus) [inline client/src/pages/Payment/CreditCard.tsx:46]: It is a good practice to add 'alt' attributes for accessibility reasons :)

**Review by rajivtitus — APPROVED** [review]
Excellent work with this! 💯
- Great use of MUI and React best practices 👍🏽

File: client/src/pages/Payment/CreditCard.tsx (hunk 7 above)
Comment (fetahokey) [reply client/src/pages/Payment/CreditCard.tsx]: done ✔️

**Comment by fetahokey** [comment]
> Excellent work with this! 💯
> 
> * Great use of MUI and React best practices 👍🏽

Thanks so much for your feedback

**Pull request merged**